sdk/python/kfp/local/executor_output_utils.py

```diff
@@ -0,0 +1,213 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for reading and processing the ExecutorOutput message."""
+import json
+import os
+from typing import Any, Dict, List, Union
+
+from google.protobuf import json_format
+from google.protobuf import struct_pb2
+from kfp import dsl
+from kfp.compiler import pipeline_spec_builder
+from kfp.dsl import executor
+from kfp.pipeline_spec import pipeline_spec_pb2
+
+
+def load_executor_output(
+        executor_output_path: str) -> pipeline_spec_pb2.ExecutorOutput:
+    """Loads the ExecutorOutput message from a path."""
+    executor_output = pipeline_spec_pb2.ExecutorOutput()
+    with open(executor_output_path) as f:
+        json_format.Parse(f.read(), executor_output)
+    return executor_output
+
+
+def cast_ints_as_floats_to_ints(
+    outputs_dict: Dict[str, Any],
+    outputs_typed_int: List[str],
+) -> Dict[str, Any]:
+    """Casts output fields that are typed as NUMBER_INTEGER to a Python int.
+
+    This is required, since google.protobuf.struct_pb2.Value uses
+    number_value to represent both floats and ints. When converting
+    struct_pb2.Value to a dict/json, int will be upcast to float, even
+    if the component output specifies int.
+    """
+    for float_output_key in outputs_typed_int:
+        outputs_dict[float_output_key] = int(outputs_dict[float_output_key])
+    return outputs_dict
+
+
+def get_outputs_from_executor_output(
+    executor_output: pipeline_spec_pb2.ExecutorOutput,
+    executor_input: pipeline_spec_pb2.ExecutorInput,
+    component_spec: pipeline_spec_pb2.ComponentSpec,
+) -> Dict[str, Any]:
+    """Obtains a dictionary of output key to output value from several messages
+    corresponding to the executed a component/task."""
+    executor_output = add_type_to_executor_output(
+        executor_input=executor_input,
+        executor_output=executor_output,
+    )
+
+    # merge any parameter outputs written using dsl.OutputPath with the rest of ExecutorOutput
+    executor_output = merge_dsl_output_file_parameters_to_executor_output(
+        executor_input=executor_input,
+        executor_output=executor_output,
+        component_spec=component_spec,
+    )
+
+    # collect parameter outputs from executor output
+    output_parameters = {
+        param_name: pb2_value_to_python(value)
+        for param_name, value in executor_output.parameter_values.items()
+    }
+    # process the special case of protobuf ints
```

Comment:
This is related to [this previous comment](https://github.com/kubeflow/pipelines/pull/10302#discussion_r1428394031). The current implementation in this PR is the correct way to handle this, IMO.